### PR TITLE
Implement ducks pattern, update tests

### PR DESCRIPTION
**Pull request by nikolasleblanc**
Based off this proposal: https://github.com/erikras/ducks-modular-redux and the work done by the Kagen Air team.

Removes individual action, constants and reducer files, and consolidates them into their respective duck file. 

Updated the tests to accommodate this change. Important to note that now that our actions are inside our duck files, our coverage threshold is expecting our action creators to be tested. I added tests for the synchronous action creators and got our coverage back up to where it should be, but still need to config an expected actions, mock store setup to test the async action creators.

### CODE REVIEW

File: src/ducks/counter.test.js [hunk 1]
```diff
@@ -1,11 +1,31 @@
 import assert from 'assert';
 import fireAction from '../utils/fire-action';
 import counterReducer from './counter';
-import { INCREMENT_COUNTER, DECREMENT_COUNTER } from '../constants';
+import {
+  INCREMENT_COUNTER,
+  DECREMENT_COUNTER,
+  increment,
+  decrement,
+} from './counter';
```
Comment (kentchau) [inline src/ducks/counter.test.js]: I would go as far importing the module with a name space alias `import * as counter from './counter'` - helps with scaling

-H

Comment (nikolasleblanc) [inline src/ducks/counter.test.js]: Yeah good idea.

**Comment by nikolasleblanc** [comment]
Setup namespacing on test imports re @kentchau's suggestion.

**Comment by SethDavenport** [comment]
@bertrandk @e-schultz thoughts on this pattern?

**Comment by bertrandk** [comment]
I haven't thought too much about it (so prepare a few grains of salt) but I do find the pattern curious given the problems I thought Redux was meant to solve.

The whole point of Redux (and Flux) is to decouple actions and reducers. Outside of trivial applications, the expected case is to have a bunch of uncoordinated reducers handling one action and a bunch of different actions being handled by one reducer. This is how the architecture scales. Large teams can work on overlapping features without conflict. As the app grows, state mutation logic can be kept separate even if it is caused by the same action (i.e. a similar number of user interactions do more as features are added). Reducers and action creators are not semantically one-to-one, which is why we don't typically co-locate them. Constants are also kept separately as a consequence of this.

Sure, decoupling sacrifices cohesion but that's not new. It's the explicit trade-off being made here. This looks nice and is convenient when the project starts but I don't know that we want to optimize for that at the detriment of allowing us to easily evolve codebases as requirements change.

Otherwise, if you somehow know your action creators are always going to be ~~glorified~~ simple setters...this looks fine to me (but it's a little questionable if you needed Redux at all). 

Anyways, I'm curious to hear what other people's experiences (or impressions) are. I understand it saves devs the effort of jumping between three files but does it do anything to reduce overall complexity or make adapting to changing product requirements simpler?

**Comment by SethDavenport** [comment]
Hmm. Ok @bertrandk has convinced me.  I think we'll leave it for now, unless @nikolasleblanc feels really strongly about it.

**Comment by nikolasleblanc** [comment]
Nah, @bertrandk convinced me too. :)

**Pull request closed without merging**